packages/controllers/src/multichain/MultiChainController.ts

```diff
@@ -0,0 +1,453 @@
+    throw new Error('Invalid account ID.');
+  }
+
+  const match = ACCOUNT_ID_REGEX.exec(accountId);
+  if (!match?.groups) {
+    throw new Error('Invalid account ID.');
+  }
+
+  return {
+    address: match.groups.accountAddress,
+    chainId: match.groups.chainId as ChainId,
+    chain: {
+      namespace: match.groups.namespace,
+      reference: match.groups.reference,
+    },
+  };
+}
+
+// TODO(ritave): Support for legacy ethereum operations, not just snaps
+export class MultiChainController extends BaseController<
+  typeof controllerName,
+  MultiChainControllerState,
+  MultiChainControllerMessenger
+> {
+  private notify: Notify;
+
+  constructor({ messenger, notify }: MultiChainControllerArgs) {
+    super({
+      messenger,
+      metadata: {
+        sessions: { persist: false, anonymous: false },
+      },
+      name: controllerName,
+      state: defaultState,
+    });
+
+    this.notify = notify;
+  }
+
+  getSession(origin: string): SessionData | undefined {
+    return this.state.sessions[origin];
+  }
+
+  async closeSession(origin: string): Promise<void> {
+    const session = this.getSession(origin);
+    assert(session, new Error('No session to close'));
+
+    await this.notify(origin, { method: 'multichainHack_metamask_disconnect' });
+
+    this.update((state) => {
+      delete state.sessions[origin];
+    });
+
+    await Promise.all(
+      Object.values(session.handlingSnaps).map((snapId) =>
+        this.messagingSystem.call('SnapController:onSessionClose', snapId),
+      ),
+    );
+  }
+
+  async onConnect(
+    origin: string,
+    connection: ConnectArguments,
+  ): Promise<Session> {
+    const existingSession = this.getSession(origin);
+    if (existingSession) {
+      await this.closeSession(origin);
+    }
+
+    const snaps = await this.messagingSystem.call('SnapController:getAll');
+    const filteredSnaps = snaps.filter((snap) => snap.enabled && !snap.blocked);
+
+    const availableNamespaces = await filteredSnaps.reduce<
+      Promise<Record<SnapId, Record<NamespaceId, Namespace>>>
+    >(async (previousPromise, snap) => {
+      const acc = await previousPromise;
+      const snapNamespaces = await this.snapToNamespaces(snap);
+      if (snapNamespaces !== null) {
+        return { ...acc, [snap.id]: snapNamespaces };
+      }
+      return acc;
+    }, Promise.resolve({}));
+
+    // The magical matching algorithm specified in SIP-2
+    const namespaceToSnaps = findMatchingKeyringSnaps(
+      connection.requiredNamespaces,
+      availableNamespaces,
+    );
+
+    const permissions = await this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      origin,
+    );
+
+    const approvedNamespacesAndSnaps = Object.entries(namespaceToSnaps).reduce(
+      (acc, [namespace, snapIds]) => {
+        // If snap already is approved for use, solve conflict by using that snap
+        const approvedSnap = snapIds.find((snapId) => {
+          return (
+            permissions &&
```

Comment:
I believe permissions is always going to be an object, maybe `{}` if there are no permissions

Reply:
Type says it can also be `undefined`, hence the check.